packages/mobile/src/pincode/authentication.test.ts

```diff
@@ -82,6 +163,174 @@ describe(getPincode, () => {
+    const mockedNavigate = navigate as jest.Mock
+    mockedNavigate.mockImplementationOnce((_, params) => {
+      params.onSuccess(mockPin)
+    })
+    mockedKeychain.getGenericPassword.mockResolvedValue({
+      password: mockPin,
+      username: 'username',
+      service: 'PIN',
+      storage: 'storage',
+    })
+
+    await setPincodeWithBiometry()
+
+    expectPincodeEntered()
+    expect(mockedKeychain.setGenericPassword).toHaveBeenCalledTimes(1)
+    expect(mockedKeychain.setGenericPassword).toHaveBeenCalledWith(
+      'CELO',
+      mockPin,
+      expect.objectContaining({
+        service: 'PIN',
+        accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET,
+        authenticationType: Keychain.AUTHENTICATION_TYPE.BIOMETRICS,
+      })
+    )
+  })
+  it('should throw an error if the retrieved pin is incorrect', async () => {
+    setCachedPin(DEFAULT_CACHE_ACCOUNT, mockPin)
+    mockedKeychain.getGenericPassword.mockResolvedValue({
+      password: 'some random password',
+      username: 'username',
+      service: 'PIN',
+      storage: 'storage',
+    })
+
+    try {
+      await setPincodeWithBiometry()
+    } catch (error) {
+      expect(error).toEqual(expect.any(Error))
+    }
+  })
+})
+
+describe(updatePin, () => {
+  const oldPin = '123123'
+  // expectedPasswordHash generated from mockPin
+  const expectedPasswordHash = '9853810edb88b031bf6ac1505f5689cb423876fbeb14f7a3037c97ec4531b6ae'
+  // expectedAccountHash generated from normalizeAddress(mockAccount)
+  const expectedAccountHash = 'PASSWORD_HASH-0000000000000000000000000000000000007e57'
+  // mockEncryptedMneumonic generated using mockMneumonic, mockAccount, old pin
+  const mockEncryptedMneumonic =
+    'U2FsdGVkX19p+azxZ2jqXIUhwbCpXi9hmfrdNMMVNYe+ptnyMGDadUzXrNJmgDyfUfmI+HXjKAcEs6XVJdeuoBFP3SH4quIeBzgjemMlq4yWFQ31TrN4TofrOuUjUuXEnnDol9Ad8gQmSK/6TmXZYXuRigwDigg9UGIKKl4SzHXgwJeWMKjnP3cOaWh9iJ8M43GfEWETJYFLCGgW6hyOeAREq6bOVP25GPcXCiE1yAM='
+
+  mockedKeychain.getGenericPassword.mockImplementation((options) => {
+    if (options?.service === 'PEPPER') {
+      return Promise.resolve(mockPepper)
+    }
+    if (options?.service === 'mnemonic') {
+      return Promise.resolve({
+        username: 'some username',
+        password: mockEncryptedMneumonic,
+        service: 'some service',
+        storage: 'some string',
+      })
+    }
+    return Promise.resolve(false)
+  })
+
+  it('should update the cached pin, stored password, and store mnemonic', async () => {
+    await updatePin(mockAccount, oldPin, mockPin)
+
+    expect(getCachedPin(DEFAULT_CACHE_ACCOUNT)).toEqual(mockPin)
+    expect(mockedKeychain.setGenericPassword).toHaveBeenCalledWith(
+      'CELO',
+      expectedPasswordHash,
+      expect.objectContaining({ service: expectedAccountHash })
+    )
+    expect(mockedKeychain.setGenericPassword).toHaveBeenCalledWith(
+      'CELO',
+      expect.any(String), // TODO test that this can be decrypted correctly
```

Comment:
@jeanregisser any suggestions on how to test that the correct encrypted mnemonic is stored? the encryption doesn't resolve to a single value so i couldn't hard code this, and we're mocking the keychain so I couldn't retrieve it to test it 🥲

Reply:
Since it's a unit test, it's fine if it's not entirely real.
But can we have a deterministic string here?

No need to do it here, but testing this e2e would ensure everything is really working as expected.